drv/psc-seq-server/src/main.rs

```diff
@@ -3,48 +3,586 @@
+            // and only watch for the presence line to indicate the PSU has
+            // appeared.
+            (PsuState::NotPresent, Present::Yes, _) => {
+                let settle_deadline = now.wrapping_add(INSERT_DEBOUNCE_MS);
+                self.state = PsuState::Present(PresentState::NewlyInserted {
+                    settle_deadline,
+                });
+                // No external action required until our timer elapses.
+                None
+            }
+
+            (
+                PsuState::Present(PresentState::NewlyInserted {
+                    settle_deadline,
+                }),
+                _,
+                _,
+            ) => {
+                if settle_deadline <= now {
+                    // The PSU is still present (since the Present::No case above
+                    // didn't fire) and our deadline has elapsed. Let's treat this
+                    // as valid!
+                    self.state = PsuState::Present(PresentState::On);
+                    Some(ActionRequired::EnableMe)
+                } else {
+                    // Remain in this state.
+                    None
+                }
+            }
+
+            // yay!
+            (PsuState::Present(PresentState::On), _, Status::Good) => None,
+
+            (PsuState::Present(PresentState::On), _, Status::NotGood) => {
+                // The PSU appears to have pulled the OK signal into the "not
+                // OK" state to indicate an internal fault!
+
+                let turn_on_deadline = now.wrapping_add(FAULT_OFF_MS);
+                self.state = PsuState::Present(PresentState::Faulted {
+                    turn_on_deadline,
+                });
+                Some(ActionRequired::DisableMe {
+                    attempt_snapshot: true,
+                })
+            }
+
+            (
+                PsuState::Present(PresentState::Faulted { turn_on_deadline }),
+                _,
+                ok,
+            ) => {
+                if turn_on_deadline <= now {
+                    match ok {
+                        Status::NotGood => {
+                            // Do it all again then, I guess.
+                            let turn_on_deadline =
+                                now.wrapping_add(FAULT_OFF_MS);
+                            self.state =
+                                PsuState::Present(PresentState::Faulted {
+                                    turn_on_deadline,
+                                });
+                            Some(ActionRequired::NoteStickyFault)
+                        }
+                        Status::Good => {
+                            self.state = PsuState::Present(PresentState::On);
+                            Some(ActionRequired::EnableMe)
+                        }
+                    }
+                } else {
+                    // Remain in this state.
+                    None
+                }
+            }
```

Comment:
take it or leave it: I kind of feel like this might be a little bit easier to follow if we pulled these out into separate match arms? not a huge deal though, it's a marginal improvement at best.
```suggestion
            // Do it all again then, I guess.
            (
                PsuState::Present(PresentState::Faulted { turn_on_deadline }),
                _,
                Status::NotGood,
            ) if turn_on_deadline <= now => {
                let turn_on_deadline =
                    now.wrapping_add(FAULT_OFF_MS);
                self.state =
                    PsuState::Present(PresentState::Faulted {
                        turn_on_deadline,
                    });
                Some(ActionRequired::NoteStickyFault)
            },

            // The PSU has indicated that it is once again good!
            (
                PsuState::Present(PresentState::Faulted { turn_on_deadline }),
                _,
                Status::Good,
            ) if turn_on_deadline <= now => {
                self.state = PsuState::Present(PresentState::On);
                Some(ActionRequired::EnableMe)
            }

            // Remain in this state.
            _ => None
```

Reply:
Hmmmm I'm not sure I feel the same on this.

I do think _all_ of these match arms were easier to read before rustfmt made them v
e
r
t
i
c
a
l
:\

Reply:
My thought process was that to me, when the match arm captured the entire condition that represented a particular state transition, rather than having additional branching logic inside it, you could view the match as a mapping from "all the inputs" to "all the outputs", which felt maybe slightly more comprehensible to _my brain_ in particular...but I think this is a matter of taste, and I'm fine with the way you have it.